move-vm-support/src/ss58_address.rs

```diff
@@ -0,0 +1,114 @@
+//! SS58 address format converter for Substrate and Move accounts.
+//! Based on the Pontem solution.
+
+use anyhow::{anyhow, ensure, Result};
+use blake2::{Blake2b512, Digest};
+use move_core_types::account_address::AccountAddress;
+
+// Substrate address prefix
+// Read more: https://docs.substrate.io/reference/address-formats/
+const SS58_PREFIX: &[u8] = b"SS58PRE";
+
+// Public key length in bytes
+const PUB_KEY_LENGTH: usize = 32;
+
+// Checksum length in bytes
+const CHECK_SUM_LEN: usize = 2;
+
+// Blake2b512 hash length in bytes
+const HASH_LEN: usize = 64;
+
+/// Convert ss58 address string to Move address structure
+/// In case if such conversion is not possible, return error.
+/// ```
+/// use move_vm_support::ss58_address::ss58_to_move_address;
+/// let substrate_address = "gkNW9pAcCHxZrnoVkhLkEQtsLsW5NWTC75cdAdxAMs9LNYCYg";
+/// let move_address = ss58_to_move_address(substrate_address).unwrap();
+/// assert_eq!(
+///    "0x8EAF04151687736326C9FEA17E25FC5287613693C912909CB226AA4794F26A48",
+///   format!("{:#X}", move_address)
+/// );
+/// ```
+pub fn ss58_to_move_address(ss58: &str) -> Result<AccountAddress> {
+    let bs58 = bs58::decode(ss58).into_vec()?;
+
+    ensure!(
+        bs58.len() > PUB_KEY_LENGTH + CHECK_SUM_LEN,
+        format!(
+            "Address length must be equal or greater than {} bytes",
+            PUB_KEY_LENGTH + CHECK_SUM_LEN
+        )
+    );
```

Comment:
> ```
> account_type: 1 || 2 bytes
> raw_account:  32 bytes
> checksum:     4 bytes
> -----------------------
> Total: 34 or 35 bytes
> ```

No no, that's wrong.
Checksum for our case is 2 bytes (not 4 bytes).
Total is 35 or 36 (not 34 or 35).

Reply:
Hmm.. maybe instead of writing our own checkers, we should just use Substrate ones:
https://docs.rs/sp-core/latest/sp_core/crypto/struct.Ss58AddressFormat.html
https://docs.rs/sp-core/latest/sp_core/crypto/trait.Ss58Codec.html

Then we'd never think what's going on inside the SS58 address 😄

Reply:
> No no, that's wrong.
Checksum in our case is 2 bytes (not 4 bytes).
Total is 35 or 36 (not 34 or 35).

Ah sorry... I meant 2 the whole time because my total was 35 & 36 at the end :smile:

Reply:
> Hmm.. maybe instead of writing our own checkers, we should just use Substrate ones:

It's still pretty simple (now that we figured it out - since the original docs are not the best source of info). At least we can make sure our code is well documented so we avoid redundant discussions

Reply:
Refactored. 
With some minor changes and fixes about clippy complaints.